In [1]:
# pip install pandas numpy keras scikit-learn
import pandas as pd 
import numpy as np 
import warnings 
warnings.filterwarnings('ignore')


# **A. Build a baseline model** 

Use the Keras library to build a neural network with the following: 
1. One hidden layer of 10 nodes, and a ReLU activation function 
2. User the `adam` optimizer and `mean squared error` as the loss function. 

Steps: 
1. Randomly split the data into training and test sets by holding 30% of the data for testing. Use `train_test_split` from Scikit-learn 
2. Train the model on the training data using `50 epochs` 
3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and actual concrete strength. Use `mean_squared_error` from Scikit-learn
4. Repeat 1-3, 50 times and create a list of 50 MSEs. 
5. Report the mean and standard deviation of the mean squared errors. 

In [2]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Cement              1030 non-null   float64
 1   Blast Furnace Slag  1030 non-null   float64
 2   Fly Ash             1030 non-null   float64
 3   Water               1030 non-null   float64
 4   Superplasticizer    1030 non-null   float64
 5   Coarse Aggregate    1030 non-null   float64
 6   Fine Aggregate      1030 non-null   float64
 7   Age                 1030 non-null   int64  
 8   Strength            1030 non-null   float64
dtypes: float64(8), int64(1)
memory usage: 72.6 KB


In [4]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

In [6]:
from sklearn.model_selection import train_test_split

y = concrete_data['Strength']
X = concrete_data[concrete_data.columns[concrete_data.columns != 'Strength']] 
test_size = 0.3
# Split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size)  
n_cols = X_train.shape[1] 
n_cols


8

In [8]:
from keras import Sequential
from keras.layers import Dense 
def regression_model(n_cols: int): 
    model = Sequential() 
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) 
    model.add(Dense(1))

    # Compile 
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model 

model = regression_model(n_cols) 
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_2 (Dense)                 │ (None, 10)             │            90 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │            11 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101 (404.00 B)

 Trainable params: 101 (404.00 B)

 Non-trainable params: 0 (0.00 B)

In [9]:
from sklearn.metrics import mean_squared_error 

def run_model():  
    # 1. Split 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size)   
    n_cols = X_train.shape[1] 
    model = regression_model(n_cols) 
    # 2. Train 
    model.fit(X_train, y_train, epochs=50, verbose=0) 
    # 3. Evaluate 
    y_predicted = model.predict(X_test)  
    mse = mean_squared_error(y_test, y_predicted) 
    
    return mse 
    

In [10]:
from tqdm import tqdm 
# 4. Repeat 50 times 
mse_values = [] 
for _ in tqdm(range(50), desc="Training model"): 
    mse_values.append(run_model()) 

mse_mean = np.mean(mse_values)
mse_std_dev = np.std(mse_values) 
print(f"MSE Mean: {mse_mean}, MSE Std. Dev: {mse_std_dev}")

Training model:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   2%|▏         | 1/50 [00:03<02:28,  3.03s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   4%|▍         | 2/50 [00:05<02:13,  2.77s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   6%|▌         | 3/50 [00:08<02:09,  2.75s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   8%|▊         | 4/50 [00:10<02:03,  2.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  10%|█         | 5/50 [00:13<01:59,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  12%|█▏        | 6/50 [00:16<01:55,  2.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  14%|█▍        | 7/50 [00:18<01:52,  2.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  16%|█▌        | 8/50 [00:21<01:49,  2.60s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  18%|█▊        | 9/50 [00:23<01:45,  2.58s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  20%|██        | 10/50 [00:26<01:43,  2.58s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  22%|██▏       | 11/50 [00:28<01:40,  2.58s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  24%|██▍       | 12/50 [00:31<01:38,  2.60s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  26%|██▌       | 13/50 [00:34<01:38,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  28%|██▊       | 14/50 [00:36<01:34,  2.63s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  30%|███       | 15/50 [00:39<01:31,  2.61s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  32%|███▏      | 16/50 [00:42<01:28,  2.61s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  34%|███▍      | 17/50 [00:44<01:25,  2.60s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  36%|███▌      | 18/50 [00:47<01:23,  2.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  38%|███▊      | 19/50 [00:50<01:21,  2.64s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  40%|████      | 20/50 [00:52<01:19,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  42%|████▏     | 21/50 [00:55<01:16,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  44%|████▍     | 22/50 [00:58<01:14,  2.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  46%|████▌     | 23/50 [01:00<01:12,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  48%|████▊     | 24/50 [01:03<01:09,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  50%|█████     | 25/50 [01:06<01:07,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  52%|█████▏    | 26/50 [01:09<01:06,  2.75s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  54%|█████▍    | 27/50 [01:11<01:02,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  56%|█████▌    | 28/50 [01:14<00:59,  2.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  58%|█████▊    | 29/50 [01:17<00:56,  2.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  60%|██████    | 30/50 [01:19<00:53,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  62%|██████▏   | 31/50 [01:22<00:50,  2.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  64%|██████▍   | 32/50 [01:25<00:48,  2.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  66%|██████▌   | 33/50 [01:27<00:45,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  68%|██████▊   | 34/50 [01:30<00:42,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  70%|███████   | 35/50 [01:33<00:40,  2.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  72%|███████▏  | 36/50 [01:35<00:37,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  74%|███████▍  | 37/50 [01:38<00:34,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  76%|███████▌  | 38/50 [01:41<00:31,  2.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  78%|███████▊  | 39/50 [01:43<00:29,  2.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  80%|████████  | 40/50 [01:46<00:27,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  82%|████████▏ | 41/50 [01:49<00:24,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  84%|████████▍ | 42/50 [01:52<00:22,  2.80s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  86%|████████▌ | 43/50 [01:55<00:20,  2.86s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  88%|████████▊ | 44/50 [01:58<00:16,  2.83s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  90%|█████████ | 45/50 [02:01<00:14,  2.90s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  92%|█████████▏| 46/50 [02:03<00:11,  2.83s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  94%|█████████▍| 47/50 [02:06<00:08,  2.78s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  96%|█████████▌| 48/50 [02:09<00:05,  2.89s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  98%|█████████▊| 49/50 [02:12<00:03,  3.00s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model: 100%|██████████| 50/50 [02:15<00:00,  2.71s/it]

MSE Mean: 312.2150544126538, MSE Std. Dev: 375.98637038907026


# **A. MSE Mean: 312.215, MSE Standard Deviation: 375.986**

# **B. Normalize the data** 

Repeat A but use a normalized version of the data. 

How does the mean of the mean squared errors compare to that of step A? 

In [11]:
def normalize(data: any): 
    mean = data.mean() 
    std = data.std() 
    return (data - mean) / std 

def run_model():  
    # 1. Split 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size)   
    n_cols = X_train.shape[1] 
    model = regression_model(n_cols) 
    # 1.1 Normalize 
    X_train_normalized = normalize(X_train)
    X_test_normalized = normalize(X_test)
    # 2. Train 
    model.fit(X_train_normalized, y_train, epochs=50, verbose=0) 
    # 3. Evaluate 
    y_predicted = model.predict(X_test_normalized)  
    mse = mean_squared_error(y_test, y_predicted) 
    
    return mse 

In [12]:
from tqdm import tqdm 
# 4. Repeat 50 times 
mse_values = [] 
for _ in tqdm(range(50), desc="Training model"): 
    mse_values.append(run_model()) 

mse_mean = np.mean(mse_values)
mse_std_dev = np.std(mse_values) 
print(f"MSE Mean: {mse_mean}, MSE Std. Dev: {mse_std_dev}")

Training model:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   2%|▏         | 1/50 [00:02<02:11,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   4%|▍         | 2/50 [00:05<02:08,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:   6%|▌         | 3/50 [00:08<02:06,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   8%|▊         | 4/50 [00:10<02:06,  2.75s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  10%|█         | 5/50 [00:13<02:01,  2.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  12%|█▏        | 6/50 [00:16<01:57,  2.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  14%|█▍        | 7/50 [00:19<01:59,  2.79s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  16%|█▌        | 8/50 [00:21<01:54,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  18%|█▊        | 9/50 [00:24<01:52,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  20%|██        | 10/50 [00:27<01:48,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  22%|██▏       | 11/50 [00:29<01:44,  2.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  24%|██▍       | 12/50 [00:32<01:41,  2.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  26%|██▌       | 13/50 [00:35<01:38,  2.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  28%|██▊       | 14/50 [00:37<01:35,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  30%|███       | 15/50 [00:40<01:32,  2.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  32%|███▏      | 16/50 [00:43<01:31,  2.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  34%|███▍      | 17/50 [00:45<01:29,  2.71s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  36%|███▌      | 18/50 [00:48<01:26,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  38%|███▊      | 19/50 [00:51<01:24,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  40%|████      | 20/50 [00:54<01:21,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  42%|████▏     | 21/50 [00:56<01:19,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  44%|████▍     | 22/50 [00:59<01:16,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  46%|████▌     | 23/50 [01:02<01:13,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  48%|████▊     | 24/50 [01:05<01:10,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  50%|█████     | 25/50 [01:07<01:08,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  52%|█████▏    | 26/50 [01:10<01:05,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  54%|█████▍    | 27/50 [01:13<01:05,  2.83s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  56%|█████▌    | 28/50 [01:16<01:01,  2.79s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  58%|█████▊    | 29/50 [01:18<00:58,  2.78s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  60%|██████    | 30/50 [01:21<00:55,  2.76s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  62%|██████▏   | 31/50 [01:24<00:52,  2.75s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  64%|██████▍   | 32/50 [01:27<00:49,  2.76s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  66%|██████▌   | 33/50 [01:29<00:46,  2.75s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  68%|██████▊   | 34/50 [01:32<00:43,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  70%|███████   | 35/50 [01:35<00:40,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  72%|███████▏  | 36/50 [01:38<00:38,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  74%|███████▍  | 37/50 [01:40<00:35,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  76%|███████▌  | 38/50 [01:43<00:32,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  78%|███████▊  | 39/50 [01:46<00:29,  2.71s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  80%|████████  | 40/50 [01:48<00:27,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  82%|████████▏ | 41/50 [01:51<00:24,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  84%|████████▍ | 42/50 [01:54<00:21,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  86%|████████▌ | 43/50 [01:57<00:19,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  88%|████████▊ | 44/50 [01:59<00:16,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  90%|█████████ | 45/50 [02:02<00:13,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  92%|█████████▏| 46/50 [02:05<00:10,  2.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  94%|█████████▍| 47/50 [02:08<00:08,  2.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  96%|█████████▌| 48/50 [02:10<00:05,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  98%|█████████▊| 49/50 [02:13<00:02,  2.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model: 100%|██████████| 50/50 [02:16<00:00,  2.73s/it]

MSE Mean: 349.6364545362458, MSE Std. Dev: 109.88859881641301


# **B. MSE Mean: 349.636 MSE Std. Dev: 109.885** 

Observations: 
1. After normalizing, the mean of the MSEs remained relatively close to the mean prior to normalizing. 
2. Key difference lies in Standard Deviation: After normalizing, results became more consistent, with a standard deviation of 109.888, while a non-normalized run generated a standard deviation of 375.986

# **C. Repeat part B but use 100 epochs** 

In [15]:

def normalize(data: any): 
    mean = data.mean() 
    std = data.std() 
    return (data - mean) / std 

def run_model():  
    # 1. Split 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size)   
    n_cols = X_train.shape[1] 
    model = regression_model(n_cols) 
    # 1.1 Normalize 
    X_train_normalized = normalize(X_train)
    X_test_normalized = normalize(X_test)
    # 2. Train 
    model.fit(X_train_normalized, y_train, epochs=100, verbose=0) 
    # 3. Evaluate 
    y_predicted = model.predict(X_test_normalized)  
    mse = mean_squared_error(y_test, y_predicted) 
    
    return mse 

In [16]:
from tqdm import tqdm 
# 4. Repeat 50 times 
mse_values = [] 
for _ in tqdm(range(50), desc="Training model"): 
    mse_values.append(run_model()) 

mse_mean = np.mean(mse_values)
mse_std_dev = np.std(mse_values) 
print(f"MSE Mean: {mse_mean}, MSE Std. Dev: {mse_std_dev}")

Training model:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   2%|▏         | 1/50 [00:04<03:47,  4.64s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   4%|▍         | 2/50 [00:09<03:41,  4.61s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   6%|▌         | 3/50 [00:13<03:34,  4.57s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:   8%|▊         | 4/50 [00:18<03:30,  4.57s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  10%|█         | 5/50 [00:22<03:24,  4.55s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  12%|█▏        | 6/50 [00:27<03:19,  4.53s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  14%|█▍        | 7/50 [00:31<03:14,  4.52s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  16%|█▌        | 8/50 [00:36<03:09,  4.52s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  18%|█▊        | 9/50 [00:40<03:05,  4.52s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  20%|██        | 10/50 [00:45<03:02,  4.56s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  22%|██▏       | 11/50 [00:50<03:00,  4.63s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  24%|██▍       | 12/50 [00:55<02:56,  4.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  26%|██▌       | 13/50 [00:59<02:53,  4.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  28%|██▊       | 14/50 [01:04<02:51,  4.76s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  30%|███       | 15/50 [01:09<02:45,  4.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  32%|███▏      | 16/50 [01:14<02:41,  4.76s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  34%|███▍      | 17/50 [01:18<02:36,  4.74s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  36%|███▌      | 18/50 [01:23<02:31,  4.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  38%|███▊      | 19/50 [01:28<02:26,  4.71s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  40%|████      | 20/50 [01:33<02:21,  4.73s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  42%|████▏     | 21/50 [01:37<02:16,  4.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  44%|████▍     | 22/50 [01:42<02:11,  4.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  46%|████▌     | 23/50 [01:47<02:06,  4.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  48%|████▊     | 24/50 [01:51<02:01,  4.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  50%|█████     | 25/50 [01:56<01:56,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  52%|█████▏    | 26/50 [02:00<01:51,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  54%|█████▍    | 27/50 [02:06<01:51,  4.85s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  56%|█████▌    | 28/50 [02:10<01:45,  4.80s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  58%|█████▊    | 29/50 [02:15<01:40,  4.81s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  60%|██████    | 30/50 [02:20<01:35,  4.76s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  62%|██████▏   | 31/50 [02:25<01:29,  4.72s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  64%|██████▍   | 32/50 [02:29<01:24,  4.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  66%|██████▌   | 33/50 [02:34<01:19,  4.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  68%|██████▊   | 34/50 [02:39<01:14,  4.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  70%|███████   | 35/50 [02:43<01:09,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  72%|███████▏  | 36/50 [02:48<01:05,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  74%|███████▍  | 37/50 [02:52<01:00,  4.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  76%|███████▌  | 38/50 [02:57<00:55,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  78%|███████▊  | 39/50 [03:02<00:51,  4.65s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  80%|████████  | 40/50 [03:07<00:46,  4.70s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  82%|████████▏ | 41/50 [03:11<00:42,  4.67s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  84%|████████▍ | 42/50 [03:16<00:37,  4.66s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  86%|████████▌ | 43/50 [03:20<00:32,  4.64s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  88%|████████▊ | 44/50 [03:25<00:27,  4.63s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  90%|█████████ | 45/50 [03:30<00:23,  4.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  92%|█████████▏| 46/50 [03:34<00:18,  4.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  94%|█████████▍| 47/50 [03:39<00:13,  4.64s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  96%|█████████▌| 48/50 [03:43<00:09,  4.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model:  98%|█████████▊| 49/50 [03:48<00:04,  4.62s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


Training model: 100%|██████████| 50/50 [03:53<00:00,  4.66s/it]

MSE Mean: 163.35939414348553, MSE Std. Dev: 16.035126259665358


# **C. MSE Mean: 163.359 MSE Std. Dev: 16.035** 

Observations: 
1. Increasing the epochs generated a considerably lower MSE to 163.359 from the previous 349.636
2. This was also able to significantly bring down the standard deviation to 16.03

# **D. Increase the number of hidden layers** 

Repeat part B but use a neural network with the following: 
1. 3 Hidden Layers, 10 Nodes, ReLU 

Compare with step B 

In [21]:
def regression_model(n_cols: int): 
    model = Sequential() 
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) 
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))

    # Compile 
    model.compile(optimizer='adam', loss='mean_squared_error') 
    return model 


def normalize(data: any): 
    mean = data.mean() 
    std = data.std() 
    return (data - mean) / std 

def run_model():  
    # 1. Split 
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_size)   
    n_cols = X_train.shape[1] 
    model = regression_model(n_cols) 
    # 1.1 Normalize 
    X_train_normalized = normalize(X_train)
    X_test_normalized = normalize(X_test)
    # 2. Train 
    model.fit(X_train_normalized, y_train, epochs=50, verbose=0) 
    # 3. Evaluate 
    y_predicted = model.predict(X_test_normalized)  
    mse = mean_squared_error(y_test, y_predicted) 
    
    return mse 

In [22]:
from tqdm import tqdm 
# 4. Repeat 50 times 
mse_values = [] 
for _ in tqdm(range(50), desc="Training model"): 
    mse_values.append(run_model()) 

mse_mean = np.mean(mse_values)
mse_std_dev = np.std(mse_values) 
print(f"MSE Mean: {mse_mean}, MSE Std. Dev: {mse_std_dev}")

Training model:   0%|          | 0/50 [00:00<?, ?it/s]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:   2%|▏         | 1/50 [00:03<02:46,  3.39s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:   4%|▍         | 2/50 [00:06<02:33,  3.20s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:   6%|▌         | 3/50 [00:09<02:28,  3.16s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:   8%|▊         | 4/50 [00:12<02:24,  3.14s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  10%|█         | 5/50 [00:15<02:20,  3.12s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  12%|█▏        | 6/50 [00:18<02:16,  3.10s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step


Training model:  14%|█▍        | 7/50 [00:21<02:13,  3.11s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 


Training model:  16%|█▌        | 8/50 [00:25<02:10,  3.11s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  18%|█▊        | 9/50 [00:28<02:07,  3.10s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  20%|██        | 10/50 [00:31<02:05,  3.14s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  22%|██▏       | 11/50 [00:35<02:12,  3.41s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  24%|██▍       | 12/50 [00:38<02:07,  3.35s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  26%|██▌       | 13/50 [00:41<02:02,  3.32s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  28%|██▊       | 14/50 [00:45<01:58,  3.30s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  30%|███       | 15/50 [00:48<01:55,  3.31s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  32%|███▏      | 16/50 [00:51<01:51,  3.28s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  34%|███▍      | 17/50 [00:54<01:47,  3.27s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  36%|███▌      | 18/50 [00:58<01:44,  3.26s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  38%|███▊      | 19/50 [01:01<01:40,  3.26s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  40%|████      | 20/50 [01:04<01:37,  3.25s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  42%|████▏     | 21/50 [01:07<01:34,  3.24s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  44%|████▍     | 22/50 [01:11<01:31,  3.26s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  46%|████▌     | 23/50 [01:14<01:27,  3.25s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  48%|████▊     | 24/50 [01:17<01:24,  3.25s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  50%|█████     | 25/50 [01:20<01:21,  3.26s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  52%|█████▏    | 26/50 [01:24<01:19,  3.30s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  54%|█████▍    | 27/50 [01:27<01:15,  3.28s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step


Training model:  56%|█████▌    | 28/50 [01:30<01:11,  3.27s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  58%|█████▊    | 29/50 [01:34<01:08,  3.28s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  60%|██████    | 30/50 [01:37<01:05,  3.27s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step 


Training model:  62%|██████▏   | 31/50 [01:40<01:01,  3.26s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  64%|██████▍   | 32/50 [01:43<00:59,  3.30s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step


Training model:  66%|██████▌   | 33/50 [01:47<00:57,  3.40s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step


Training model:  68%|██████▊   | 34/50 [01:53<01:07,  4.19s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step


Training model:  70%|███████   | 35/50 [01:59<01:10,  4.69s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  72%|███████▏  | 36/50 [02:03<01:01,  4.40s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  74%|███████▍  | 37/50 [02:08<00:59,  4.56s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step


Training model:  76%|███████▌  | 38/50 [02:12<00:52,  4.41s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  78%|███████▊  | 39/50 [02:15<00:45,  4.14s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  80%|████████  | 40/50 [02:18<00:38,  3.86s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  82%|████████▏ | 41/50 [02:22<00:33,  3.68s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  84%|████████▍ | 42/50 [02:25<00:28,  3.53s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  86%|████████▌ | 43/50 [02:28<00:23,  3.41s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  88%|████████▊ | 44/50 [02:31<00:20,  3.34s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step


Training model:  90%|█████████ | 45/50 [02:34<00:16,  3.29s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  92%|█████████▏| 46/50 [02:37<00:12,  3.24s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  94%|█████████▍| 47/50 [02:41<00:09,  3.22s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  96%|█████████▌| 48/50 [02:44<00:06,  3.22s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model:  98%|█████████▊| 49/50 [02:47<00:03,  3.20s/it]

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step 


Training model: 100%|██████████| 50/50 [02:50<00:00,  3.41s/it]

MSE Mean: 131.99406150301792, MSE Std. Dev: 15.7755440357826


# **D. MSE Mean: 131.994 MSE Std. Dev: 15.775** 

Observations: 
1. Increasing the number of hidden layers exhibits similar behavior to increasing number of epochs: Improved MSE from ~300 to just ~100. 
2. Results also became more consistent

### **Conclusion: Increasing number of layers and number of epochs improves model performance, at the cost of training time and computational resources**